Composite: make items tabbable if active element gets removed

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

May fix https://github.com/WordPress/gutenberg/issues/65283

This PR tweaks the behavior of the `Composite` component so that the composite widget can continue to receive keyboard focus even when its active ID doesn't correspond to an element rendered in the DOM.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The fact that `Composite` can end up in a situation where the whole widget is not focusable is bad for assistive technology, and we should try to fix it.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Following [a suggestion](https://github.com/ariakit/ariakit/issues/4129) from the Ariakit maintainers, this PR makes all composite items tabbable if the element associated with the active ID is not connected to the store.

In this case, when the focus is moved back to the composite widget, the items will be able to receive focus. As soon as one item receives focus, all composite items will stop being tabbable and revert back to the standard behavior of a composite widget.

## Review suggestions

Review commit-by-commit. The actual fix is applied in the first commit. The rest of the commits are accessories to the main fix.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

In Storybook:
- Load the "Remove active item" example
- Focus the composite widget, and with the arrow keys move focus on the third item
- Tab to the "Toggle third item button" and click it
- Tab back to the composite widget: the second item should receive focus and become active
- Moving arrow keys should work as expected in the composite widget

<details>

<summary>Edit: it looks like the UI has changed and it's impossible to reproduce the bug in the editor as of now</summary>

In the editor:
- Follow the testing instructions indicated in https://github.com/WordPress/gutenberg/issues/65283
- When clicking the "randomize colors" button, tab back to the color palette.
- The last item in the palette should receive keyboard focus
- Moving arrow keys should work as expected in the composite widget

</details>

## Screenshots or screencast <!-- if applicable -->

| Before (trunk) | After (this PR) |
|---|---|
| <video src="https://github.com/user-attachments/assets/d7535e1c-d70e-4ee2-a7e3-bbdadeec1e32" /> | <video src="https://github.com/user-attachments/assets/d91a8172-e2d8-460f-967d-cc9b65bff417" /> |

ciampo committed:
Composite: make items tabbable when the active element is disconnected

ciampo committed:
Add unit test

ciampo commented on packages/components/src/composite/item.tsx:
```diff
@@ -20,9 +20,24 @@ export const CompositeItem = forwardRef<
+	const storeViaProps = props.store as Ariakit.CompositeStore;
+	const store = storeViaProps ?? ( context.store as Ariakit.CompositeStore );
+
+	// If the active item is not connected, Composite may end up in a state
+	// where none of the items are tabbable. In this case, we force all items to
+	// be tabbable, so that as soon as an item received focus, it becomes active
+	// and Composite goes back to working as expected.
+	const tabbable = Ariakit.useStoreState( store, ( state ) => {
+		return ! store?.item( state?.activeId )?.element?.isConnected;
+	} );
+
```
This is the actual fix

diegohaz replied:
Just a heads up: it's probably better to check if `activeId` is not `null` beforehand. See https://github.com/ariakit/ariakit/issues/4129#issuecomment-2367928139

ciampo replied:
Thank you, updated!

ciampo commented on packages/components/src/composite/stories/index.story.tsx:
I'll remove this story before merging

ciampo commented on packages/components/src/composite/legacy/test/index.tsx:
While reviewing some unit test failures, I realised that `renderAndValidate` is an async function but is never `await`ed. If folks think this change is worth applying, I'll do so in a separate PR.

ciampo replied:
https://github.com/WordPress/gutenberg/pull/65823

ciampo commented on packages/components/src/composite/group.tsx:
```diff
@@ -20,11 +20,11 @@ export const CompositeGroup = forwardRef<
-		/>
-	);
+
+	// @ts-expect-error The store prop in undocumented and only used by the
+	// legacy compat layer.
+	const storeViaProps = props.store as Ariakit.CompositeStore;
+	const store = storeViaProps ?? ( context.store as Ariakit.CompositeStore );
```
Reading the store passed via `props` is a fix to how the legacy compat layer works. I believe that this is a bug that's been around since we wrote the legacy compat layer, but it was surfaced by the changes in this PR. If folks agree with it, I'll apply it in a separate PR.

ciampo replied:
https://github.com/WordPress/gutenberg/pull/65821

ciampo commented on packages/components/src/composite/test/index.tsx:
We should probably add a suite of tests similar to the legacy version. I can do that in a follow-up PR

tyxla replied:
Sounds like a plan

ciampo replied:
https://github.com/WordPress/gutenberg/pull/65952

tyxla commented on packages/components/src/composite/item.tsx:
```diff
@@ -20,9 +20,27 @@ export const CompositeItem = forwardRef<
+	// where none of the items are tabbable. In this case, we force all items to
+	// be tabbable, so that as soon as an item received focus, it becomes active
+	// and Composite goes back to working as expected.
+	const tabbable = Ariakit.useStoreState( store, ( state ) => {
+		return (
+			state?.activeId !== null &&
+			! store?.item( state?.activeId )?.element?.isConnected
+		);
+	} );
```
Good catch 👍 

I wonder if this is something that Ariakit should resolve internally?

ciampo replied:
I already discussed it in https://github.com/ariakit/ariakit/issues/4129. I think Ariakit maintainers are cautious about implementing such a fix directly in the library, and instead asked if we wanted to try the fix out in userland first

tyxla commented on packages/components/src/composite/stories/index.story.tsx:
```diff
@@ -353,3 +353,28 @@ const Fill = ( { children } ) => {
 	},
 };
+
+export const RemoveActiveItem: StoryObj< typeof Composite > = {
```
Was going to say this story doesn't make much sense to keep, but I see you created it solely for testing and plan to remove it - thanks!

tyxla commented on packages/components/src/composite/test/index.tsx:
```diff
@@ -0,0 +1,123 @@
+ */
+import { Composite } from '..';
+
+// This is necessary because of how Ariakit calculates page up and
+// page down. Without this, nothing has a height, and so paging up
+// and down doesn't behave as expected in tests.
```
Ah, jsdom strikes again. Makes you wonder if this kind of tests isn't better suited for e2e. I guess not, because we're not really testing anything related to calculating the height.

ciampo replied:
Not singularly for this test, but in general we way need to consider this. The closer our environment is to the real browser, the more confidence we get from the tests

tyxla commented on packages/components/src/composite/test/index.tsx:
```diff
@@ -0,0 +1,123 @@
+}
+
+describe( 'Composite', () => {
+	it( 'should remain focussable even when there is no elements in the DOM associated with the currently active ID', async () => {
```
Nit: let's use the US English version like for anything else we do in code:

```suggestion
	it( 'should remain focusable even when there is no elements in the DOM associated with the currently active ID', async () => {
```

ciampo replied:
There's also a grammar error that I'll fix ("there IS no elements" -> "there ARE no elements")

tyxla commented on packages/components/src/composite/test/index.tsx:
```diff
@@ -0,0 +1,123 @@
+	);
+}
+
+describe( 'Composite', () => {
```
Thanks for adding a comprehensive test 👍

tyxla reviewed (commented):
Looks good, works well 👍 

Thanks for adding tests and a special storybook story to manually test it in 🙇‍♂️ 

Not approving just yet since we'll need to clean up after merging #65823 and #65821.

ciampo committed:
Better test name

ciampo committed:
CHANGELOG

ciampo commented:
Rebased to include changes from https://github.com/WordPress/gutenberg/pull/65823 and https://github.com/WordPress/gutenberg/pull/65821 (which were merged in the meantime), and addressed other pending feedback

@tyxla this is ready for another round of review

tyxla reviewed (approved):
LGTM 🚀

ciampo commented:
I'll enable backporting once https://github.com/WordPress/gutenberg/pull/65853 gets merged

diegohaz commented:
@ciampo Is there anything noteworthy about this change? Have you encountered any issues?

afercia commented:
@ciampo @diegohaz I haven't tested the progress on this fix since I reported it iin https://github.com/WordPress/gutenberg/issues/65283

Do I get it correctly that the fix is:
- When there's no element referenced by the activeId, all the items become focusable.
- As soon as one of the items receives focus, the original behavior is restored with only one item tabbable (tabindex=0) and the other ones not tabbable (tabindex=-1). Arrows navigation will work as expected.

If that's correct, it seems to me it's a reasonable and effective fix.
